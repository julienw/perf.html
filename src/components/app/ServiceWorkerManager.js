/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import React, { PureComponent } from 'react';
import explicitConnect from '../../utils/connect';
import { assertExhaustiveCheck } from '../../utils/flow';

import { getDataSource } from '../../selectors/url-state';
import { getView } from '../../selectors/app';
import { getSymbolicationStatus } from '../../selectors/profile';

import type { ConnectedProps } from '../../utils/connect';
import type { DataSource } from '../../types/actions';
import type { Phase, SymbolicationStatus } from '../../types/state';

import './ServiceWorkerManager.css';

type StateProps = {|
  +dataSource: DataSource,
  +phase: Phase,
  +symbolicationStatus: SymbolicationStatus,
|};
type Props = ConnectedProps<{||}, StateProps, {||}>;

type InstallStatus = 'pending' | 'ready' | 'idle';
type State = {|
  installStatus: InstallStatus,
  isNoticeDisplayed: boolean,
|};

/**
 * This component is responsible for installing and updating the service worker,
 * possibly displaying an information notice to the user to reload the page
 * whenever a new version is ready and we're in a state that would accept a
 * reload.
 *
 * Here are some assumptions:
 * - The browser checks for an update (in background) only at startup.
 * - By the time we have a symbolicated profile displayed, everything that
 *   should be downloaded is already downloaded.
 *
 * This means that we can force the new service worker at that time: this
 * shouldn't disturb other open tabs that are fully loaded already, nor this
 * current tab that is also fully loaded.
 *
 * If we don't force the new service worker, then, to update, the user will have
 * to close all tabs first. This is not very nice, especially that it's common
 * that our users have several tabs open.
 *
 * In the future we could decide to force the new service worker only when the
 * user clicks the reload button. This might be necessary if we need to download
 * new content after the profile viewer is fully loaded.
 * But be wary that this breaks the F5 capability: the user would _have_ to
 * click the button to update (or close all tabs like said above).
 */

class ServiceWorkerManager extends PureComponent<Props, State> {
  state = {
    installStatus: 'idle',
    isNoticeDisplayed: false,
  };

  _isProfileLoadedAndReady(): boolean {
    const { phase, symbolicationStatus } = this.props;

    if (phase !== 'DATA_LOADED') {
      // Note we don't use a switch for the phase because it has a lot of
      // different values and won't likely change often. Hopefully this comment
      // won't age badly.
      return false;
    }

    switch (symbolicationStatus) {
      case 'DONE':
        return true;
      case 'SYMBOLICATING':
        return false;
      default:
        throw assertExhaustiveCheck(symbolicationStatus);
    }
  }

  _shouldUpdateServiceWorker(): boolean {
    const { dataSource } = this.props;

    switch (dataSource) {
      case 'from-file':
      case 'from-addon':
        // We should not propose to reload the page for these data sources,
        // because we'd lose the data.
        return false;
      case 'none':
        // But for these data sources it should be fine.
        return true;
      case 'public':
      case 'from-url':
      case 'compare':
      case 'local':
        // Before updating the service worker, we need to make sure the profile
        // is ready -- which means we don't need to download anything more.
        return this._isProfileLoadedAndReady();
      default:
        throw assertExhaustiveCheck(dataSource);
    }
  }

  _installServiceWorker() {
    const runtime = require('offline-plugin/runtime');
    runtime.install({
      onInstalled: () => {
        console.log('[ServiceWorker] App is ready for offline usage!');
      },
      onUpdating: () => {
        console.log(
          '[ServiceWorker] An update has been found and the browser is downloading the new assets.'
        );
      },
      onUpdateReady: () => {
        console.log(
          '[ServiceWorker] We have downloaded the new assets and we are ready to go.'
        );
        if (this._shouldUpdateServiceWorker()) {
          runtime.applyUpdate();
        }
        this.setState({ installStatus: 'pending' });
      },
      onUpdated: () => {
        console.log(
          '[ServiceWorker] The new version of the application has been enabled.'
        );
        this.setState({ installStatus: 'ready', isNoticeDisplayed: true });
      },
      onUpdateFailed: () => {
        console.log(
          '[ServiceWorker] We failed to update the application for an unknown reason.'
        );
      },
    });
  }

  componentDidMount() {
    if (process.env.NODE_ENV === 'production') {
      this._installServiceWorker();
    }
  }

  componentDidUpdate() {
    const { phase } = this.props;
    const { installStatus } = this.state;

    if (installStatus !== 'idle' && phase === 'FATAL_ERROR') {
      // If we got a fatal error and a new version of the application is
      // available, let's try to reload automatically, as this might fix the
      // fatal error.
      const runtime = require('offline-plugin/runtime');
      runtime.applyUpdate();
      this.reloadPage();
    } else if (
      installStatus === 'pending' &&
      this._shouldUpdateServiceWorker()
    ) {
      // We may have received a service worker update before we were ready to
      // update (for example: we were still symbolicating). We can apply the
      // new service worker now. When the new service worker will be enabled,
      // the notice will be displayed.
      const runtime = require('offline-plugin/runtime');
      runtime.applyUpdate();
    }
  }

  reloadPage() {
    window.location.reload();
  }

  _onCloseNotice = () => {
    this.setState({ isNoticeDisplayed: false });
  };

  render() {
    const { isNoticeDisplayed } = this.state;

    if (!isNoticeDisplayed) {
      return null;
    }

    if (!this._shouldUpdateServiceWorker()) {
      return null;
    }

    return (
      <div className="serviceworker-ready-notice-wrapper">
        {/* We use the wrapper to horizontally center the notice */}
        <div className="photon-message-bar serviceworker-ready-notice">
          A new version of the application has been downloaded and is ready to
          use.
          <button
            className="photon-button photon-button-micro photon-message-bar-action-button"
            type="button"
            onClick={this.reloadPage}
          >
            Reload the application
          </button>
          <button
            aria-label="Hide the reload notice"
            title="Hide the reload notice"
            className="photon-button photon-message-bar-close-button"
            type="button"
            onClick={this._onCloseNotice}
          />
        </div>
      </div>
    );
  }
}

export default explicitConnect<{||}, StateProps, {||}>({
  mapStateToProps: state => ({
    phase: getView(state).phase,
    dataSource: getDataSource(state),
    symbolicationStatus: getSymbolicationStatus(state),
  }),
  component: ServiceWorkerManager,
});
