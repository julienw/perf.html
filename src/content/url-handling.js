/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import queryString from 'query-string';
import { stringifyRangeFilters, parseRangeFilters } from './range-filters';
import { stringifyCallTreeFilters, parseCallTreeFilters } from './call-tree-filters';
import type { URLState } from './reducers/types';

// {
//   // general:
//   dataSource: 'from-addon', 'from-file', 'local', 'public',
//   hash: '' or 'aoeurschsaoeuch',
//   selectedTab: 'summary' or 'calltree' or ...,
//   rangeFilters: [] or [{ start, end }, ...],
//   selectedThread: 0 or 1 or ...,
//
//   // only when selectedTab === 'calltree':
//   callTreeSearchString: '' or '::RunScript' or ...,
//   callTreeFilters: [[], [{type:'prefix', matchJSOnly:true, prefixFuncs:[1,3,7]}, {}, ...], ...], // one per thread
//   jsOnly: false or true,
//   invertCallstack: false or true,
// }

function dataSourceDirs(urlState: URLState) {
  const { dataSource } = urlState;
  switch (dataSource) {
    case 'from-addon':
      return ['from-addon'];
    case 'from-file':
      return ['from-file'];
    case 'local':
      return ['local', urlState.hash];
    case 'public':
      return ['public', urlState.hash];
    default:
      return [];
  }
}

export function urlFromState(urlState: URLState) {
  const { dataSource } = urlState;
  if (dataSource === 'none') {
    return '/';
  }
  const pathname = '/' + [
    ...dataSourceDirs(urlState),
    urlState.selectedTab,
  ].join('/') + '/';

  // Start with the query parameters that are shown regardless of the active tab.
  const query: Object = {
    range: stringifyRangeFilters(urlState.rangeFilters) || undefined,
    thread: `${urlState.selectedThread}`,
  };

  if (process.env.NODE_ENV === 'development') {
    /* eslint-disable camelcase */
    query.react_perf = null;
    /* eslint-enable camelcase */
  }

  // Depending on which tab is active, also show tab-specific query parameters.
  switch (urlState.selectedTab) {
    case 'calltree':
      query.search = urlState.callTreeSearchString || undefined;
      query.invertCallstack = urlState.invertCallstack ? null : undefined;
      query.implementation = urlState.implementation === 'combined'
        ? undefined
        : urlState.implementation;
      query.callTreeFilters = stringifyCallTreeFilters(urlState.callTreeFilters[urlState.selectedThread]) || undefined;
      break;
    case 'timeline':
      query.search = urlState.callTreeSearchString || undefined;
      query.invertCallstack = urlState.invertCallstack ? null : undefined;
      query.hidePlatformDetails = urlState.hidePlatformDetails ? null : undefined;
      break;
  }
  const qString = queryString.stringify(query);
  return pathname + (qString ? '?' + qString : '');
}

export function stateFromCurrentLocation(): URLState {
  console.log('location is', window.location);
  const pathname = window.location.pathname;
  const qString = window.location.search.substr(1);
  const hash = window.location.hash;
  const query = queryString.parse(qString);

  if (pathname === '/') {
    const legacyQuery = Object.assign({}, query, queryString.parse(hash));
    if ('report' in legacyQuery) {
      if ('filter' in legacyQuery) {
        const filters = JSON.parse(legacyQuery.filter);
        // We can't convert these parameters to the new URL parameters here
        // because they're relative to different things - the legacy range
        // filters were relative to profile.meta.startTime, and the new
        // rangeFilters param is relative to
        // getTimeRangeIncludingAllThreads(profile).start.
        // So we stuff this information into a global here, and then later,
        // once we have the profile, we convert that information into URL params
        // again. This is not pretty.
        window.legacyRangeFilters =
          filters.filter(f => f.type === 'RangeSampleFilter').map(({ start, end }) => ({ start, end }));
      }
      return {
        dataSource: 'public',
        hash: legacyQuery.report,
        selectedTab: 'calltree',
        rangeFilters: [],
        selectedThread: 0,
        callTreeSearchString: '',
        callTreeFilters: {},
        implementation: 'combined',
        invertCallstack: false,
        hidePlatformDetails: false,
      };
    }
  }

  const dirs = pathname.split('/').filter(d => d);
  const dataSource = dirs[0] || 'none';
  if (!['none', 'from-addon', 'from-file', 'local', 'public'].includes(dataSource)) {
    throw new Error('unexpected data source');
  }
  const needHash = ['local', 'public'].includes(dataSource);
  const selectedThread = query.thread !== undefined ? +query.thread : 0;

  let implementation = 'combined';
  if (query.implementation === 'js' || query.implementation === 'cpp') {
    implementation = query.implementation;
  } else if (query.jsOnly !== undefined) {
    // Support the old URL structure that had a jsOnly flag.
    implementation = 'js';
  }

  return {
    dataSource,
    hash: needHash ? dirs[1] : '',
    selectedTab: (needHash ? dirs[2] : dirs[1]) || 'calltree',
    rangeFilters: query.range ? parseRangeFilters(query.range) : [],
    selectedThread: selectedThread,
    callTreeSearchString: query.search || '',
    callTreeFilters: {
      [selectedThread]: query.callTreeFilters ? parseCallTreeFilters(query.callTreeFilters) : [],
    },
    implementation,
    invertCallstack: query.invertCallstack !== undefined,
    hidePlatformDetails: query.hidePlatformDetails !== undefined,
  };
}
