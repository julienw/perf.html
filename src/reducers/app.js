/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { combineReducers } from 'redux';

import { getSelectedTab } from './url-state';
import { tabSlugs } from '../app-logic/tabs-handling';

import type { TabSlug } from '../app-logic/tabs-handling';
import type { Action } from '../types/store';
import type {
  State,
  AppState,
  AppViewState,
  IsSidebarOpenPerPanelState,
  Reducer,
} from '../types/reducers';

function view(
  state: AppViewState = { phase: 'INITIALIZING' },
  action: Action
): AppViewState {
  if (state.phase === 'DATA_LOADED') {
    // Let's not come back at another phase if we're already displaying a profile
    return state;
  }

  switch (action.type) {
    case 'TEMPORARY_ERROR_RECEIVING_PROFILE_FROM_STORE':
    case 'TEMPORARY_ERROR_RECEIVING_PROFILE_FROM_URL':
    case 'TEMPORARY_ERROR_RECEIVING_PROFILE_FROM_ADDON':
      return {
        phase: 'INITIALIZING',
        additionalData: {
          message: action.error.message,
          attempt: action.error.attempt,
        },
      };
    case 'ERROR_RECEIVING_PROFILE_FROM_FILE':
    case 'FATAL_ERROR_RECEIVING_PROFILE_FROM_ADDON':
    case 'FATAL_ERROR_RECEIVING_PROFILE_FROM_STORE':
    case 'FATAL_ERROR_RECEIVING_PROFILE_FROM_URL':
      return { phase: 'FATAL_ERROR', error: action.error };
    case 'WAITING_FOR_PROFILE_FROM_ADDON':
      return { phase: 'INITIALIZING' };
    case 'ROUTE_NOT_FOUND':
      return { phase: 'ROUTE_NOT_FOUND' };
    case 'RECEIVE_ZIP_FILE':
    case 'VIEW_PROFILE':
      return { phase: 'DATA_LOADED' };
    default:
      return state;
  }
}

function isUrlSetupDone(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'URL_SETUP_DONE':
      return true;
    default:
      return state;
  }
}

function hasZoomedViaMousewheel(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'HAS_ZOOMED_VIA_MOUSEWHEEL': {
      return true;
    }
    default:
      return state;
  }
}

function isSidebarOpenPerPanel(
  state: IsSidebarOpenPerPanelState,
  action: Action
): IsSidebarOpenPerPanelState {
  if (state === undefined) {
    state = {};
    tabSlugs.forEach(tabSlug => (state[tabSlug] = false));
  }

  switch (action.type) {
    case 'CHANGE_SIDEBAR_OPEN_STATE': {
      const { tab, isOpen } = action;
      // Due to how this action will be dispatched we'll always have the value
      // changed so we don't need the performance optimization of checking the
      // stored value against the new value.
      return {
        ...state,
        [tab]: isOpen,
      };
    }
    default:
      return state;
  }
}

/**
 * The panels that make up the timeline, details view, and sidebar can all change
 * their sizes depending on the state that is fed to them. In order to control
 * the invalidations of this sizing information, provide a "generation" value that
 * increases monotonically for any change that potentially changes the sizing of
 * any of the panels. This provides a mechanism for subscribing components to
 * deterministically update their sizing correctly.
 */
function panelLayoutGeneration(state: number = 0, action: Action): number {
  switch (action.type) {
    case 'INCREMENT_PANEL_LAYOUT_GENERATION':
    // Sidebar: (fallthrough)
    case 'CHANGE_SIDEBAR_OPEN_STATE':
    // Timeline: (fallthrough)
    case 'HIDE_GLOBAL_TRACK':
    case 'SHOW_GLOBAL_TRACK':
    case 'ISOLATE_PROCESS':
    case 'ISOLATE_PROCESS_MAIN_THREAD':
    case 'HIDE_LOCAL_TRACK':
    case 'SHOW_LOCAL_TRACK':
    case 'ISOLATE_LOCAL_TRACK':
    // Committed range changes: (fallthrough)
    case 'COMMIT_RANGE':
    case 'POP_COMMITTED_RANGES':
      return state + 1;
    default:
      return state;
  }
}

/**
 * Clicking on tracks can switch between different tabs. This piece of state holds
 * on to the last relevant thread-based tab that was viewed. This makes the UX nicer
 * for when a user clicks on a Network track, and gets taken to the Network
 * panel, then clicks on a thread or process track. With this state we can smoothly
 * transition them back to the panel they were using.
 */
function lastVisibleThreadTabSlug(
  state: TabSlug | null = null,
  action: Action
): TabSlug | null {
  switch (action.type) {
    case 'SELECT_TRACK':
    case 'CHANGE_SELECTED_TAB':
      if (action.selectedTab !== 'network-chart') {
        return action.selectedTab;
      }
      return state;
    default:
      return state;
  }
}

const appStateReducer: Reducer<AppState> = combineReducers({
  view,
  isUrlSetupDone,
  hasZoomedViaMousewheel,
  isSidebarOpenPerPanel,
  panelLayoutGeneration,
  lastVisibleThreadTabSlug,
});

export default appStateReducer;

export const getApp = (state: State): AppState => state.app;
export const getView = (state: State): AppViewState => getApp(state).view;
export const getIsUrlSetupDone = (state: State): boolean =>
  getApp(state).isUrlSetupDone;
export const getHasZoomedViaMousewheel = (state: State): boolean => {
  return getApp(state).hasZoomedViaMousewheel;
};
export const getIsSidebarOpen = (state: State): boolean =>
  getApp(state).isSidebarOpenPerPanel[getSelectedTab(state)];
export const getPanelLayoutGeneration = (state: State) =>
  getApp(state).panelLayoutGeneration;
export const getLastVisibleThreadTabSlug = (state: State) =>
  getApp(state).lastVisibleThreadTabSlug;
