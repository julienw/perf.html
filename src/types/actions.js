/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { CallTree } from '../profile-logic/call-tree';
import type {
  Profile,
  Thread,
  ThreadIndex,
  IndexIntoMarkersTable,
  IndexIntoFuncTable,
} from './profile';
import type { IndexIntoCallNodeTable, CallNodeTable } from './profile-derived';
import type { GetLabel } from '../profile-logic/labeling-strategies';
import type { GetCategory } from '../profile-logic/color-categories';
import type { TemporaryError } from '../utils/errors';
import type { Transform } from './transforms';

export type DataSource =
  | 'none'
  | 'from-file'
  | 'from-addon'
  | 'local'
  | 'public'
  | 'from-url';
export type ProfileSelection =
  | { hasSelection: false, isModifying: false }
  | {
      hasSelection: true,
      isModifying: boolean,
      selectionStart: number,
      selectionEnd: number,
    };
export type FuncToFuncMap = Map<IndexIntoFuncTable, IndexIntoFuncTable>;
export type FunctionsUpdatePerThread = {
  [id: ThreadIndex]: {
    oldFuncToNewFuncMap: FuncToFuncMap,
    funcIndices: IndexIntoFuncTable[],
    funcNames: string[],
  },
};

export type RequestedLib = { debugName: string, breakpadId: string };
export type ImplementationFilter = 'combined' | 'js' | 'cpp';
export type TabSlug =
  | 'calltree'
  | 'stack-chart'
  | 'marker-chart'
  | 'marker-table'
  | 'flame-graph';

type ProfileAction =
  | { type: 'ROUTE_NOT_FOUND', url: string }
  | { type: 'CHANGE_THREAD_ORDER', threadOrder: ThreadIndex[] }
  | {
      type: 'HIDE_THREAD',
      threadIndex: ThreadIndex,
      hiddenThreads: ThreadIndex[],
      threadOrder: ThreadIndex[],
    }
  | { type: 'SHOW_THREAD', threadIndex: ThreadIndex }
  | {
      type: 'ISOLATE_THREAD',
      hiddenThreadIndexes: ThreadIndex[],
      isolatedThreadIndex: ThreadIndex,
    }
  | {
      type: 'ASSIGN_TASK_TRACER_NAMES',
      addressIndices: number[],
      symbolNames: string[],
    }
  | {
      type: 'CHANGE_SELECTED_CALL_NODE',
      threadIndex: ThreadIndex,
      selectedCallNodeIndex: IndexIntoCallNodeTable | null,
      callNodeAncestors: IndexIntoCallNodeTable[],
    }
  | {
      type: 'FOCUS_CALL_TREE',
    }
  | {
      type: 'CHANGE_EXPANDED_CALL_NODES',
      threadIndex: ThreadIndex,
      expandedCallNodeIndexes: Set<IndexIntoCallNodeTable>,
    }
  | {
      type: 'CHANGE_SELECTED_MARKER',
      threadIndex: ThreadIndex,
      selectedMarker: IndexIntoMarkersTable | -1,
    }
  | { type: 'UPDATE_PROFILE_SELECTION', selection: ProfileSelection }
  | { type: 'CHANGE_TAB_ORDER', tabOrder: number[] };

type ReceiveProfileAction =
  | {
      type: 'COALESCED_FUNCTIONS_UPDATE',
      functionsUpdatePerThread: FunctionsUpdatePerThread,
    }
  | { type: 'DONE_SYMBOLICATING' }
  | { type: 'ERROR_RECEIVING_PROFILE_FROM_FILE', error: Error }
  | {
      type: 'TEMPORARY_ERROR_RECEIVING_PROFILE_FROM_ADDON',
      error: TemporaryError,
    }
  | { type: 'FATAL_ERROR_RECEIVING_PROFILE_FROM_ADDON', error: Error }
  | {
      type: 'TEMPORARY_ERROR_RECEIVING_PROFILE_FROM_STORE',
      error: TemporaryError,
    }
  | {
      type: 'TEMPORARY_ERROR_RECEIVING_PROFILE_FROM_URL',
      error: TemporaryError,
    }
  | { type: 'FATAL_ERROR_RECEIVING_PROFILE_FROM_STORE', error: Error }
  | { type: 'FATAL_ERROR_RECEIVING_PROFILE_FROM_URL', error: Error }
  | { type: 'RECEIVE_PROFILE_FROM_ADDON', profile: Profile }
  | { type: 'RECEIVE_PROFILE_FROM_FILE', profile: Profile }
  | { type: 'RECEIVE_PROFILE_FROM_STORE', profile: Profile }
  | { type: 'RECEIVE_PROFILE_FROM_URL', profile: Profile }
  | { type: 'REQUESTING_SYMBOL_TABLE', requestedLib: RequestedLib }
  | { type: 'RECEIVED_SYMBOL_TABLE_REPLY', requestedLib: RequestedLib }
  | { type: 'START_SYMBOLICATING' }
  | { type: 'WAITING_FOR_PROFILE_FROM_ADDON' }
  | { type: 'WAITING_FOR_PROFILE_FROM_STORE' }
  | { type: 'WAITING_FOR_PROFILE_FROM_URL' };

type StackChartAction =
  | { type: 'CHANGE_STACK_CHART_COLOR_STRATEGY', getCategory: GetCategory }
  | { type: 'CHANGE_STACK_CHART_LABELING_STRATEGY', getLabel: GetLabel }
  | { type: 'HAS_ZOOMED_VIA_MOUSEWHEEL' };

type UrlEnhancerAction =
  | { type: '@@urlenhancer/urlSetupDone' }
  | { type: '@@urlenhancer/updateUrlState', urlState: any };

type UrlStateAction =
  | { type: 'WAITING_FOR_PROFILE_FROM_FILE' }
  | { type: 'PROFILE_PUBLISHED', hash: string }
  | { type: 'CHANGE_SELECTED_TAB', selectedTab: TabSlug }
  | { type: 'ADD_RANGE_FILTER', start: number, end: number }
  | { type: 'POP_RANGE_FILTERS', firstRemovedFilterIndex: number }
  | { type: 'CHANGE_SELECTED_THREAD', selectedThread: ThreadIndex }
  | { type: 'CHANGE_RIGHT_CLICKED_THREAD', selectedThread: ThreadIndex }
  | { type: 'CHANGE_CALL_TREE_SEARCH_STRING', searchString: string }
  | {
      type: 'ADD_TRANSFORM_TO_STACK',
      threadIndex: ThreadIndex,
      transform: Transform,
      transformedThread: Thread,
    }
  | {
      type: 'POP_TRANSFORMS_FROM_STACK',
      threadIndex: ThreadIndex,
      firstRemovedFilterIndex: number,
    }
  | {
      type: 'CHANGE_IMPLEMENTATION_FILTER',
      implementation: ImplementationFilter,
      threadIndex: ThreadIndex,
      transformedThread: Thread,
      previousImplementation: ImplementationFilter,
      implementation: ImplementationFilter,
    }
  | {|
      type: 'CHANGE_INVERT_CALLSTACK',
      invertCallstack: boolean,
      callTree: CallTree,
      callNodeTable: CallNodeTable,
      selectedThreadIndex: ThreadIndex,
    |}
  | { type: 'CHANGE_HIDE_PLATFORM_DETAILS', hidePlatformDetails: boolean }
  | { type: 'CHANGE_MARKER_SEARCH_STRING', searchString: string };

type IconsAction =
  | { type: 'ICON_HAS_LOADED', icon: string }
  | { type: 'ICON_IN_ERROR', icon: string };

export type Action =
  | ProfileAction
  | ReceiveProfileAction
  | StackChartAction
  | UrlEnhancerAction
  | UrlStateAction
  | IconsAction;
