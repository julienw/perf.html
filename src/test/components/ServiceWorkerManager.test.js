/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import React from 'react';
import { Provider } from 'react-redux';
import { render, fireEvent } from 'react-testing-library';
import serviceWorkerRuntime from 'offline-plugin/runtime';

import ServiceWorkerManager from '../../components/app/ServiceWorkerManager';
import { stateFromLocation } from '../../app-logic/url-handling';
import { updateUrlState } from '../../actions/app';
import {
  fatalError,
  viewProfile,
  startSymbolicating,
  doneSymbolicating,
} from '../../actions/receive-profile';

import { blankStore } from '../fixtures/stores';
import { getProfileFromTextSamples } from '../fixtures/profiles/processed-profile';

// Mock the offline plugin library.
jest.mock('offline-plugin/runtime', () => ({
  install: jest.fn(),
  applyUpdate: jest.fn(),
}));

function _getSimpleProfile() {
  return getProfileFromTextSamples('A').profile;
}

describe('app/ServiceWorkerManager', () => {
  // Opt out of Flow checking for this variable because we're doing
  // unconventional things with it.
  let nativeLocation: any;
  beforeEach(() => {
    // Because of how window.location is implemented in browsers and jsdom, we
    // can't easily spy on `window.location.reload`. That's why we replace the
    // full property 'location' instead.

    nativeLocation = Object.getOwnPropertyDescriptor(window, 'location');

    // It seems node v8 doesn't let us change the value unless we delete it before.
    delete window.location;
    // $FlowExpectError because the value we pass isn't a proper Location object.
    Object.defineProperty(window, 'location', {
      value: { reload: jest.fn() },
      writable: true,
      configurable: true,
    });
  });

  afterEach(() => {
    process.env.NODE_ENV = 'development';

    Object.defineProperty(window, 'location', nativeLocation);
    nativeLocation = null;
  });

  function setup() {
    jest.spyOn(console, 'log').mockImplementation(() => {});

    function navigateToStoreLoadingPage() {
      const newUrlState = stateFromLocation({
        pathname: '/public/ThisIsAFakeHash/calltree',
        search: '',
        hash: '',
      });
      store.dispatch(updateUrlState(newUrlState));
    }

    function navigateToAddonLoadingPage() {
      const newUrlState = stateFromLocation({
        pathname: '/from-addon/',
        search: '',
        hash: '',
      });
      store.dispatch(updateUrlState(newUrlState));
    }

    const store = blankStore();

    const renderResult = render(
      <Provider store={store}>
        <ServiceWorkerManager />
      </Provider>
    );

    return {
      ...renderResult,
      navigateToStoreLoadingPage,
      navigateToAddonLoadingPage,
      dispatch: store.dispatch,
    };
  }

  it('does not register a service worker in the development environment', () => {
    setup();
    expect(serviceWorkerRuntime.install).not.toHaveBeenCalled();
  });

  describe('in the home, with the `none` datasource', () => {
    it('calls the appropriate functions to update the service worker', () => {
      process.env.NODE_ENV = 'production';

      setup();
      expect(serviceWorkerRuntime.install).toHaveBeenCalled();

      const installOptions = serviceWorkerRuntime.install.mock.calls[0][0];
      installOptions.onUpdating();
      expect(serviceWorkerRuntime.applyUpdate).not.toHaveBeenCalled();
      installOptions.onUpdateReady();
      expect(serviceWorkerRuntime.applyUpdate).toHaveBeenCalled();
      installOptions.onUpdated();
      installOptions.onInstalled();
      expect(console.log).toHaveBeenCalledTimes(4);
    });

    it('shows a notice when the SW is updated, and the user can close it', () => {
      process.env.NODE_ENV = 'production';

      const { container, getByLabelText, getByText } = setup();

      const installOptions = serviceWorkerRuntime.install.mock.calls[0][0];
      installOptions.onUpdated();
      expect(console.log).toHaveBeenCalled();
      expect(container.firstChild).toMatchSnapshot();

      const closeButton = getByLabelText(/hide/i);
      fireEvent.click(closeButton);
      expect(container.firstChild).toBe(null);

      // Getting a new update should display the notice again
      installOptions.onUpdated();
      const reloadButton = getByText(/reload/i);
      fireEvent.click(reloadButton);
      expect(window.location.reload).toHaveBeenCalledTimes(1);
    });
  });

  describe('with the `public` datasource', () => {
    it(`doesn't show a notice until a profile is fully loaded`, async () => {
      process.env.NODE_ENV = 'production';

      const { navigateToStoreLoadingPage, container, dispatch } = setup();

      navigateToStoreLoadingPage();

      const installOptions = serviceWorkerRuntime.install.mock.calls[0][0];

      installOptions.onUpdateReady();
      expect(serviceWorkerRuntime.applyUpdate).not.toHaveBeenCalled();
      expect(container.firstChild).toBe(null);

      await dispatch(viewProfile(_getSimpleProfile()));
      expect(serviceWorkerRuntime.applyUpdate).toHaveBeenCalled();

      installOptions.onUpdated();
      expect(container.firstChild).not.toBe(null);
    });

    it(`doesn't show a notice until we're done symbolicating`, async () => {
      process.env.NODE_ENV = 'production';

      const { navigateToStoreLoadingPage, dispatch, container } = setup();

      navigateToStoreLoadingPage();
      await dispatch(viewProfile(_getSimpleProfile()));
      dispatch(startSymbolicating());

      const installOptions = serviceWorkerRuntime.install.mock.calls[0][0];

      installOptions.onUpdateReady();
      expect(serviceWorkerRuntime.applyUpdate).not.toHaveBeenCalled();
      expect(container.firstChild).toBe(null);

      dispatch(doneSymbolicating());
      expect(serviceWorkerRuntime.applyUpdate).toHaveBeenCalled();

      installOptions.onUpdated();
      expect(container.firstChild).not.toBe(null);
    });
  });

  it('does not show a notice with the `from-addon` datasource', () => {
    process.env.NODE_ENV = 'production';

    const { navigateToAddonLoadingPage, container } = setup();
    navigateToAddonLoadingPage();

    const installOptions = serviceWorkerRuntime.install.mock.calls[0][0];
    installOptions.onUpdateReady();
    installOptions.onUpdated();
    expect(serviceWorkerRuntime.applyUpdate).not.toHaveBeenCalled();
    expect(container.firstChild).toBe(null);
  });

  describe('automatic reloading', () => {
    it('reloads the application automatically if there is an error and there is a pending version', () => {
      process.env.NODE_ENV = 'production';

      const { dispatch } = setup();

      const installOptions = serviceWorkerRuntime.install.mock.calls[0][0];

      dispatch(fatalError(new Error('Error while loading profile')));
      expect(window.location.reload).not.toHaveBeenCalled();
      expect(serviceWorkerRuntime.applyUpdate).not.toHaveBeenCalled();
      installOptions.onUpdateReady();
      expect(serviceWorkerRuntime.applyUpdate).toHaveBeenCalled();
      expect(window.location.reload).toHaveBeenCalled();
    });

    it('reloads the application automatically if there is an error and there is a ready version happening then', () => {
      process.env.NODE_ENV = 'production';

      const { dispatch } = setup();
      const installOptions = serviceWorkerRuntime.install.mock.calls[0][0];

      dispatch(fatalError(new Error('Error while loading profile')));
      expect(window.location.reload).not.toHaveBeenCalled();
      installOptions.onUpdated();
      expect(window.location.reload).toHaveBeenCalled();
    });

    it('reloads the application automatically if there is an error and a new version has been notified before', () => {
      process.env.NODE_ENV = 'production';

      const { dispatch } = setup();
      const installOptions = serviceWorkerRuntime.install.mock.calls[0][0];

      installOptions.onUpdated();
      expect(window.location.reload).not.toHaveBeenCalled();
      dispatch(fatalError(new Error('Error while loading profile')));
      expect(window.location.reload).toHaveBeenCalled();
    });
  });
});
