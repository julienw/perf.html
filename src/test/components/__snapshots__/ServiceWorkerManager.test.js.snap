// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/ServiceWorkerManager in the home, with the \`none\` datasource shows a notice when the SW is updated, and the user can close it 1`] = `
<div
  class="serviceworker-ready-notice-wrapper"
>
  <div
    class="photon-message-bar serviceworker-ready-notice"
  >
    A new version of the application has been downloaded and is ready to use.
    <button
      class="photon-button photon-button-micro photon-message-bar-action-button"
      type="button"
    >
      Reload the application
    </button>
    <button
      aria-label="Hide the reload notice"
      class="photon-button photon-message-bar-close-button"
      title="Hide the reload notice"
      type="button"
    />
  </div>
</div>
`;
